src/main/java/org/opensearch/neuralsearch/search/query/HybridCollectorManager.java

```diff
@@ -150,66 +148,62 @@ private Collector getHybridQueryCollector() {
+        DocValueFormat[] docValueFormats = getSortValueFormats(sortAndFormats);
+        for (Collector collector : hybridSearchCollectors) {
+            TopDocsAndMaxScore topDocsAndMaxScore = getTopDocsAndAndMaxScore(collector, docValueFormats);
+            results.add((QuerySearchResult result) -> reduceCollectorResults(result, topDocsAndMaxScore, docValueFormats));
         }
+        return results;
+    }
 
-        // TODO: Cater the fix for the Bug https://github.com/opensearch-project/neural-search/issues/799
-        if (!hybridSortedTopDocCollectors.isEmpty()) {
-            HybridTopFieldDocSortCollector hybridSortedTopScoreDocCollector = hybridSortedTopDocCollectors.stream()
-                .findFirst()
-                .orElseThrow(() -> new IllegalStateException("cannot collect results of hybrid search query"));
-
-            List<TopFieldDocs> topFieldDocs = hybridSortedTopScoreDocCollector.topDocs();
-            long maxTotalHits = hybridSortedTopScoreDocCollector.getTotalHits();
-            float maxScore = hybridSortedTopScoreDocCollector.getMaxScore();
-
-            TopDocs newTopDocs = getNewTopFieldDocs(
-                getTotalHits(this.trackTotalHitsUpTo, topFieldDocs, isSingleShard, maxTotalHits),
+    private TopDocsAndMaxScore getTopDocsAndAndMaxScore(final Collector collector, final DocValueFormat[] docValueFormats) {
+        float maxScore;
+        TopDocs newTopDocs;
+        if (docValueFormats != null) {
+            HybridTopFieldDocSortCollector hybridTopFieldDocSortCollector = (HybridTopFieldDocSortCollector) collector;
+            List<TopFieldDocs> topFieldDocs = hybridTopFieldDocSortCollector.topDocs();
+            maxScore = hybridTopFieldDocSortCollector.getMaxScore();
+            newTopDocs = getNewTopFieldDocs(
+                getTotalHits(this.trackTotalHitsUpTo, topFieldDocs, hybridTopFieldDocSortCollector.getTotalHits()),
                 topFieldDocs,
                 sortAndFormats.sort.getSort()
             );
-            TopDocsAndMaxScore topDocsAndMaxScore = new TopDocsAndMaxScore(newTopDocs, maxScore);
-            return (QuerySearchResult result) -> { result.topDocs(topDocsAndMaxScore, getSortValueFormats(sortAndFormats)); };
+        } else {
+            HybridTopScoreDocCollector hybridTopScoreDocCollector = (HybridTopScoreDocCollector) collector;
+            List<TopDocs> topDocs = hybridTopScoreDocCollector.topDocs();
+            maxScore = hybridTopScoreDocCollector.getMaxScore();
+            newTopDocs = getNewTopDocs(getTotalHits(this.trackTotalHitsUpTo, topDocs, hybridTopScoreDocCollector.getTotalHits()), topDocs);
         }
+        return new TopDocsAndMaxScore(newTopDocs, maxScore);
+    }
 
-        throw new IllegalStateException("cannot collect results of hybrid search query, there are no proper score collectors");
+    private List<Collector> getHybridSearchCollectors(final Collection<Collector> collectors) {
```

Comment:
I like that we return list of abstract collectors. Can you check if it's possible to create an interface common of all hybrid query related collectors and use it instread of Collector interface? If that's possible it should also help in other places where we want to abstract code related to collector object, but have to explicitly case to one of two hybrid specific collector types